lib/Dialect/AIEVec/Transforms/ConvertVectorToAIEVec.cpp

```diff
@@ -89,6 +113,115 @@ struct UPDOpEffectiveAccessSizeAnalysis {
 //===----------------------------------------------------------------------===//
 // Lowering patterns
 //===----------------------------------------------------------------------===//
+// This pattern replaces `vector.fma` with `aievec.mac_elem`.
+struct LowerVectorFMAToAIEVecFMAElem
+    : public OpConversionPattern<vector::FMAOp> {
+  using OpConversionPattern<vector::FMAOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(vector::FMAOp fmaOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    VectorType resultType = fmaOp.getResult().getType().cast<VectorType>();
+
+    if (!resultType.getElementType().isa<IntegerType>())
+      return failure();
+
+    IntegerType resultElType = resultType.getElementType().cast<IntegerType>();
+    unsigned resultElWidth = resultElType.getWidth();
+    unsigned laneSize = getVectorLaneSize(resultType);
+
+    if ((laneSize != 32 || resultElWidth != 16) &&
+        (laneSize != 16 || resultElWidth != 32))
+      return failure();
+
+    Value acc = adaptor.getAcc();
+    Type accType = getVectorOpDestType(acc.getType().cast<VectorType>(), true);
+
+    if (!writesToAccumulator(acc.getDefiningOp())) {
+      acc = rewriter.create<xilinx::aievec::UPSOp>(fmaOp->getLoc(), accType,
+                                                   acc, 0);
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\n\nCreated UPS op " << acc << " to move the output of "
+                 << fmaOp << " into accumulator");
+    }
+
+    auto bcastOp =
+        dyn_cast<vector::BroadcastOp>(adaptor.getRhs().getDefiningOp());
+    Value lhs = adaptor.getLhs();
+
+    if (!bcastOp) {
+      bcastOp = dyn_cast<vector::BroadcastOp>(adaptor.getLhs().getDefiningOp());
+      lhs = adaptor.getRhs();
+      if (!bcastOp)
+        return failure();
+    }
+
+    auto extOp =
+        dyn_cast<vector::ExtractOp>(bcastOp.getSource().getDefiningOp());
+    if (!extOp)
+      return failure();
+
+    auto newRhs = extOp.getVector();
+    auto pos = extOp.getPosition();
+
+    Value rhsVal = rewriter.create<aievec::BroadcastOp>(
+        fmaOp->getLoc(), resultType, newRhs,
+        cast<IntegerAttr>(pos[0]).getInt());
+
+    auto fmaElemOp = rewriter.create<aievec::FMAElemOp>(
+        fmaOp->getLoc(), lhs, rhsVal, acc, /*fmsub=*/false);
+
+    for (auto user : fmaOp->getUsers()) {
```

Comment:
IMO, this pattern should replace the vector.fma with a elem.mac_elem if the elem type of the mac_elem and the fma are the same.
Otherwise, if the elem type of the mac_elem is wider than that of the vector.fma, it should replace the vector.fma with a mac_elem followed by an SRS.
The pattern should not check the users of the vector.fma; this is not relevant.